src/Infrastructure.VS/ProjectDirectoryProvider.cs

```diff
@@ -0,0 +1,94 @@
+﻿/*
+ * SonarLint for Visual Studio
+ * Copyright (C) 2016-2021 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace SonarLint.VisualStudio.Infrastructure.VS
+{
+    public interface IProjectDirectoryProvider
+    {
+        /// <summary>
+        /// Returns full path of the project containing the given source file
+        /// </summary>
+        string GetProjectDirectory(string sourceFilePath);
+    }
+
+    [Export(typeof(IProjectDirectoryProvider))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class ProjectDirectoryProvider : IProjectDirectoryProvider
+    {
+        /// <summary>
+        /// See https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.__vspropid7?view=visualstudiosdk-2019
+        /// The enum is not available in VS 2015 API.
+        /// </summary>
+        internal const int VSPROPID_IsInOpenFolderMode = -8044;
+
+        private readonly IVsHierarchyLocator vsHierarchyLocator;
+        private readonly IVsSolution vsSolution;
+
+        [ImportingConstructor]
+        public ProjectDirectoryProvider([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IVsHierarchyLocator vsHierarchyLocator)
+        {
+            vsSolution = serviceProvider.GetService(typeof(SVsSolution)) as IVsSolution;
+            this.vsHierarchyLocator = vsHierarchyLocator;
+        }
+
+        public string GetProjectDirectory(string sourceFilePath)
+        {
+            // "Open as Folder" was introduced in VS2017, so in VS2015 the hr result will be `E_NOTIMPL` and `isOpenAsFolder` will be null.
```

Comment:
Looking at this again, we're making a fairly big assumption, namely that in the "open as folder" case the supplied file path is under the folder root. That's true for the way we call it currently, but it looks odd now the interface has been separated out.

Effectively, the behaviour in the "open as folder" case is that the caller can pass in any garbage and the function will return the root folder path.
I think we should either document this behaviour on the interface, or add a check to the code.

Reply:
This is actually exactly the design problem I'm 'struggling' with right now.
Eventually the higher caller would call something like
`CompilationOptions Get(string analyzedFile)` --> and that in turn would get the compilation database, get the entry and parse it.
So in a way we do always start with a file, but strictly speaking, I'm writing a class now `CompilationDatabase Locate()` that absolutely does not need a file path.